### PR TITLE
ClientConn: Introduce GetStatus method to get the latest connection error

### LINKED ISSUE #2031: TLS failures in blocking Dial calls don't provide useful error messages
### What version of gRPC are you using?
v1.9.2, but I've verified the same issue exists at the latest release (v1.11.3).

### What version of Go are you using (`go version`)?
1.10

### What operating system (Linux, Windows, …) and version?
Linux

### What did you do?
I'm using a blocking Dial (`grpc.DialContext(..., grpc.WithBlock())`) to better account errors to the dial phase vs RPC calls. One of the backends had an unexpected CN in its TLS certificate, so the connection was failing. The error message returned from `DialContext` was only `context deadline exceeded`, and the useful errors were printed out into the info log.

I expected a blocking dial to return some indication of why it had failed, instead of only timing out. In particular, including the most recent transport error in the returned error value would be very useful for reporting connection problems.

Based on the conversation in https://github.com/grpc/grpc-go/pull/1855 it sounds like this behavior is intentional (or, at least, expected). Is there any way to either get access to the transport-layer errors, or somehow propagate them up into the returned error value?

The very useful error printed to the logs:
`W0427 21:43:33.738993   35711 clientconn.go:1167] grpc: addrConn.createTransport failed to connect to {badname.example.com 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: x509: certificate is valid for goodname.example.com, not badname.example.com". Reconnecting...`

The not very useful error returned from `DialContext`:
`W0427 21:43:33.773061   35711 handlers.go:180] Error dialing backend "badname.example.com": context deadline exceeded`

### PR DESCRIPTION
**Pull request by dfawley**
Fixes #2031 

This may be just a proof of concept, but I think we will need something like this.  A few things I noticed:

- Is the "last connection error" a better fit alongside the connectivity state (like in Java) instead of within the blockingpicker?
- We probably should either clear the latest connection error on any success (to prevent leaking the current error to a subsequent failure state), and/or be very sure it's always set freshly when the status of the channel goes to transient failure.  I would definitely prefer to do the latter.
- GetStatus() is racy because there is no lock.  However, the user's code is unavoidably racy between noticing `GetState() == TransientFailure` and doing the `GetStatus()` call, so maybe it's fine?

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -719,6 +719,16 @@ func (cc *ClientConn) GetState() connectivity.State {
...
 }
 
+// GetStatus returns the status of cc as an error produced by the status
+// package.  If the state is TransientFailure, the status error will be
+// non-nil.  Otherwise, it will always be nil.
+func (cc *ClientConn) GetStatus() error {
+	if cc.GetState() != connectivity.TransientFailure {
+		return nil
+	}
+	return status.Errorf(codes.Unavailable, "latest connection error: %v", cc.blockingpicker.connectionError())
```
Comment (menghanl) [inline clientconn.go]: What's the advantage of using status error here?
What other error codes will we return?

File: clientconn.go [hunk 2]
```diff
@@ -877,6 +887,9 @@ func (cc *ClientConn) removeAddrConn(ac *addrConn, err error) {
 		return
 	}
 	delete(cc.conns, ac)
+	if len(cc.conns) == 0 {
+		cc.blockingpicker.updateConnectionError(errors.New("no backends returned by resolver"))
```
[the remark this answers is not in the document]
Comment (dfawley) [reply clientconn.go]: Isn't this that code path?

File: clientconn.go (hunk 1 above)
Comment (dfawley) [reply clientconn.go]: It mirrors the Java API.  That one returns a status directly, but I thought "error" would be more idiomatic for us.

**Comment by dfawley** [comment]
Closing for now until I get a chance to revisit.

**Pull request closed without merging**